File: 03/hs/03.hs
#!/usr/bin/env runhaskell

go' dy dx input =
    length . filter id $ zipWith f (every dy input) [0..]
        where f x i = cycle x !! (dx*i) == '#'
              every n [] = []
              every n xs@(x:_) = x:every n (drop n xs)

go = go' 1

main = do
    input <- lines <$> readFile "input"
    putStrLn . show $ go 3 input
    putStrLn . show $ product $ map ($ input) [go 1, go 3, go 5, go 7, go' 2 1]

File: 03/rs/03.rs
use std::fs::File;
use std::io::{self, BufRead};

fn go(input: &[String], dx: usize, dy: usize) -> usize {
    input.iter().step_by(dy).enumerate()
        .filter(|(y, line)| line.chars().cycle().nth(dx*y) == Some('#')).count()
}

fn main() -> io::Result<()> {
    let f = File::open("input")?;
    let input: Vec<String> = io::BufReader::new(f).lines().flatten().collect();

    println!("{}", go(&input, 3, 1));
    println!("{}", [(1,1),(3,1),(5,1),(7,1),(1,2)]
        .iter().map(|&(dx,dy)| go(&input, dx, dy)).product::<usize>());

    Ok(())
}
